[Lens] Rank top values by custom metric

mbondyra opened the pull request:
## Summary

Fixes #133991.
<img width="331" alt="Screenshot 2022-07-04 at 10 19 59" src="https://user-images.githubusercontent.com/4283304/177113240-70871c2e-fc6f-4da5-a87e-6f5fc83061f7.png">
<img width="322" alt="Screenshot 2022-07-04 at 10 20 05" src="https://user-images.githubusercontent.com/4283304/177113244-1d0ef8ea-676c-42b1-9de3-80adfa52397d.png">
<img width="326" alt="Screenshot 2022-07-04 at 10 20 12" src="https://user-images.githubusercontent.com/4283304/177113248-a83d0e38-59a7-4142-aa9c-2d97e0d88162.png">



- [x] added new `orderBy` option: `custom`
- [x] added new param to terms column : `orderAgg`
- [x] refactored `ReferenceEditor` to also be able to edit inline columns, not just referenced ones 
- [x] refactored metrics `paramEditor` to accept also inlined columns and not referenced by columnId inside layer
Extra fixes:
- [x] Fix: `___records___` displayed as  `Records` when  moving from count to another operation type
<img width="685" alt="Screenshot 2022-06-27 at 11 13 56" src="https://user-images.githubusercontent.com/4283304/175905490-988bccb5-270a-473d-80d7-4d8ca05431bc.png">

- [x] Fix: `array` last value option should impossible to select when last value is a referenced operation.
<img width="757" alt="Screenshot 2022-06-27 at 11 15 56" src="https://user-images.githubusercontent.com/4283304/175905544-a461064f-4fe0-4304-9228-0e3b865d10ff.png">

- [x] Fix: correct all the dimension panel elements to occupy the full width for mobile, eg this one before fix:
<img width="561" alt="Screenshot 2022-06-29 at 09 43 29" src="https://user-images.githubusercontent.com/4283304/176380978-58d57ab1-8adb-45bd-b2e7-52c19568bc83.png">


### Checklist

mbondyra commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/index.ts:
```diff
@@ -646,8 +665,9 @@ interface OperationDefinitionMap<C extends BaseIndexPatternColumn, P = {}> {
-> = BaseOperationDefinitionProps<C> & OperationDefinitionMap<C, P>[Input];
+  P = {},
+  AR extends boolean = false
+> = BaseOperationDefinitionProps<C, AR> & OperationDefinitionMap<C, AR, P>[Input];
```
`AR` is passed down to `OperationDefinitionMap` and `FieldBasedOperationDefinition`, but it seems like you solved it within `BaseOperationDefinitionProps`

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/dimension_editor.tsx:
```diff
@@ -60,8 +61,8 @@ import { FieldInput } from './field_input';
 import { ParamEditorProps } from '../operations/definitions';
 import { WrappingHelpPopover } from '../help_popover';
-
-const operationPanels = getOperationDisplay();
```
Why did you pull this deeper into the functions? It seems like it's completely static anyway

mbondyra replied:
I couldn't make the tests work with the current setup :( I memoized the function so I think that solves the problem with unnecessary re-calculation. Let me know if you have a better idea.

flash1293 replied:
What test is problematic?

mbondyra replied:
It's actually only a problem for `ReferenceEditor` - the last commit I just pushed shows it - around 40 tests suites break with the same:
<img width="579" alt="Screenshot 2022-06-29 at 11 12 13" src="https://user-images.githubusercontent.com/4283304/176399383-018f1b01-ed36-4e96-8ab1-f49a0b9f014c.png">

It has something to do with importing reference editor inside the terms operation definition, but I cannot figure it out :(

mbondyra replied:
Actually, it's not just tests, Lens also doesn't load properly with: 
`TypeError: Object(...) is not a function  at ./public/indexpattern_datasource/dimension_panel/reference_editor.tsx`

flash1293 commented:
When changing a param in the reference editor on moving average, then it gets rid of the moving average:
* Moving average of percentile of bytes
* Change percentile value to 20
* Switches to 20th percentile of bytes (without moving average)

mbondyra commented:
> When changing a param in the reference editor on moving average, then it gets rid of the moving average [...]

 ✔️ Done (accidentally I passed the parent column id in the updater function instead of referencedColumnid in Dimension Editor)

mbondyra commented:
@flash1293 addressed all your comments, thanks for your review! 

 ✔️Cyclic import fixed 

> When changing a param in the reference editor on moving average, then it gets rid of the moving average [...]

 ✔️ Done (accidentally I passed the parent column id in the updater function instead of referencedColumnid in Dimension Editor)

flash1293 reviewed (approved):
Tested and couldn't find any issues - LGTM.
As it's a pretty complex PR I would appreciate another look from someone else

mbondyra commented:
@elasticmachine merge upstream

stratoula reviewed (commented):
This works great and is an amazing addition. I have only found one bug.

I have created a chart like this
<img width="1215" alt="image" src="https://user-images.githubusercontent.com/17003240/177275765-20e17d88-c6f8-4068-ba6d-aec69a371ba3.png">

I am using a custom rank by count. When I click on the chart to create a filter, the filter is not created and I see the following error on the console
<img width="305" alt="image" src="https://user-images.githubusercontent.com/17003240/177275921-06d6599a-c9f4-46d1-8146-b70680066388.png">

flash1293 commented:
Actually I think it's happening somewhere else and it might be harder to fix in general... We put the data table into the redux store (`activeData`), which causes it to get frozen, but then we pass the same data table to the filter click handler which apparently tries to mutate it. I guess it should be fixed on the app services side - passed in agg params shouldn't get mutated, that's totally unexpected

flash1293 commented:
Deep cloning the params here makes it work: https://github.com/elastic/kibana/blob/a8095ce11adcee821e10458c1ed1dc30173cc5a3/src/plugins/data/public/actions/filters/create_filters_from_value_click.ts#L94

mbondyra commented:
@stratoula and @flash1293 thank you for finding the bug and for the investigation! I've fixed it with suggested solution, @stratoula  please recheck when you have a moment :)

stratoula reviewed (approved):
Thanx Marta and Joe! Now it works fine! LGTM

MichaelMarcialis commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/shared_components/form_row.scss:
```diff
@@ -0,0 +1,3 @@
+.lnsInnerIndexPattern__inlineRow {
+  min-width: 96px;
```
Again, I hadn't considered the "Percentile ranks value" input when putting together the mockups. I think we'll need to increase the `min-width` here in order to ensure the prepended labels are the same width (or alternatively come up with a shorter label than "Percentile ranks value").

![image](https://user-images.githubusercontent.com/3884767/177376657-78c02080-5a1e-42a4-83f8-216a488a40ce.png)

```suggestion
  min-width: 144px;
```

mbondyra replied:
That unfortunately has the other effect when we don't use percentile ranks - it takes too much space... 

<img width="351" alt="Screenshot 2022-07-05 at 23 15 22" src="https://user-images.githubusercontent.com/4283304/177417849-4d02713c-66aa-4eda-99a1-ee3b43fa13c1.png">

MichaelMarcialis replied:
Yeah. What about a shorter name for "Percentile ranks value"? For example, would simply "Percentile rank" work? Then we could adjust the label min-width to be equal to the largest of the labels.

MichaelMarcialis commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/last_value.tsx:
```diff
@@ -342,41 +348,43 @@ export const lastValueOperation: OperationDefinition<
               {
                 defaultMessage:
-                  'Displays all values associated with this field in each last document.',
+                  'When you show array values, you are unable to use this field to rank Top values.',
```
```suggestion
                  'When you show array values, you are unable to use this field to rank top values.',
```

MichaelMarcialis commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/shared_components/form_row.scss:
```diff
@@ -0,0 +1,3 @@
+.lnsInnerIndexPattern__inlineRow {
```
Is this the most appropriate class name for this element? Asking because the `lnsInnerIndexPattern` based classes seem to be found in the field list, not the configuration flyout. Based on the parent element classes, I'm wondering if something like the following would be better.

```suggestion
.lnsIndexPatternDimensionEditor__labelCustomRank {
```

MichaelMarcialis reviewed (approved):
Thanks for putting this together, @mbondyra! This looks great. I'm leaving you a handful of small comments/suggestions below. Assuming those get addressed, this looks good from my perspective. Approving now so I don't hold you up.

Dosant reviewed (approved):
Thanks for the agg config change! lgtm!
